example_workspace/src/hello_world_example_app/app/src/main/java/com/ros/example/hello_ros/MainActivity.java

```diff
@@ -0,0 +1,159 @@
+package com.ros.example.hello_ros;
+
+import android.Manifest;
+import android.app.Activity;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+import android.widget.TextView;
+import android.widget.Toast;
+
+import java.lang.Runnable;
+import java.net.Inet4Address;
+import java.net.InetAddress;
+import java.net.NetworkInterface;
+import java.util.Enumeration;
+
+public class MainActivity extends Activity {
+    static {
+        System.loadLibrary("native-activity");
+    }
+
+    private final String TAG = "HELLO-WORLD-EXAMPLE";
+
+    private enum Status {
+        WAITING, RUNNING
+    }
+
+    private class RosThread implements Runnable {
+        public RosThread() {
+            Log.i(TAG, "calling __RosThread");
+            __RosThread();
+            Log.i(TAG, "ended __RosThread");
+        }
+        @Override
+        public native void run();
+        public native void stop();
+        public native boolean checkRosMaster(String masterIP, String myIP);
+        private native void __RosThread();
+    }
+
+    private RosThread mainThread;
+    private EditText masterIP;
+    private EditText myIP;
+    private Button runButton;
+    private TextView statusText;
+    private Status status;
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        Log.i(TAG, "OnCreate()");
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.logging);
+
+        masterIP = (EditText) findViewById(R.id.master_ip);
+        myIP = (EditText) findViewById(R.id.my_ip);
+        runButton = (Button) findViewById(R.id.run_button);
+        statusText = (TextView) findViewById(R.id.status);
+        status = Status.WAITING;
+
+        Log.i(TAG, "new RosThread");
+        mainThread = new RosThread();
+        Log.i(TAG, "end RosThread constructor");
+
+        runButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                runButtonCallback();
+            }
+        });
+
+        String ip = getLocalIpAddress();
+        if (ip != null) {
+            myIP.setText(ip);
+        }
+    }
+
+    @Override
+    protected void onResume() {
+        super.onResume();
+        if (status == Status.RUNNING) {
+            new Thread(mainThread).start();
```

Comment:
It's been a while that I last programmed Java. This construct looks a bit weird. From a C++ background it looks like the `start()` method would be called on a temporary `Thread` object which then gets immediately destroyed/garbage collected?

Reply:
There are many ways of firing threads in Java; this one in particular is not necessarily incorrect. Here's a [basic reference](http://tutorials.jenkov.com/java-concurrency/creating-and-starting-threads.html) to the topic.

What actually matters is that the `stop` method actually makes the thread exit its loop. In this basic example it calls `ros::shutdown`, so the main thread's loop will exit as well. I tried starting / stopping the thread multiple times and it seems to be doing its job properly: the node is killed on stop and relaunched on start.

Reply:
Thanks for the explanation. I only wondered whether the thread can be terminated earlier than when `ros::shutdown()` gets called as soon as the garbage collector kicks in. But [apparently this is not the case](https://stackoverflow.com/questions/2423284/java-thread-garbage-collected-or-not) **as long as it is running**. So the code is correct.